File: 189. Rotate Array/main.c
void reverse(int *array, int start, int end) {
    while (start < end) {
        int temp = array[start];
        array[start] = array[end];
        array[end] = temp;
        start++;
        end--;
    }
}

void rotate(int *nums, int numsSize, int k) {
    int count = k % numsSize;

    reverse(nums, 0, numsSize - 1);
    reverse(nums, 0, count - 1);
    reverse(nums, count, numsSize - 1);
}

File: 189. Rotate Array/main.rs
impl Solution {
    pub fn rotate(nums: &mut Vec<i32>, k: i32) {
        let count = k as usize % nums.len();
        nums.rotate_right(count);
    }
}

impl Solution {
    pub fn rotate(nums: &mut Vec<i32>, k: i32) {
        let index = nums.len() - k as usize % nums.len();
        nums.clone_from(
            &nums[index..]
                .iter()
                .chain(nums[..index].iter())
                .cloned()
                .collect(),
        );
    }
}

impl Solution {
    pub fn rotate(nums: &mut Vec<i32>, k: i32) {
        let index = nums.len() - k as usize % nums.len();
        nums.clone_from([nums[index..].cloned(), nums[..index].cloned()].concat());
    }
}

impl Solution {
    pub fn rotate(nums: &mut Vec<i32>, k: i32) {
        let index = nums.len() - k as usize % nums.len();
        nums.clone_from([nums[index..].cloned(), nums[..index].cloned()].concat());
    }
}

impl Solution {
    pub fn rotate(nums: &mut Vec<i32>, k: i32) {
        let index = nums.len() - k as usize % nums.len();
        nums[..index].reverse();
        nums[index..].reverse();
        nums.reverse();
    }
}
